### ISSUE TITLE
@BeforeGroups only called if group is specified explicitly

### ISSUE THREAD

**Issue by quandor** [cbeust/testng#118]
There are 3 Testclasses each belonging to a different group. Each Test has a method annotated with @BeforeGroups. Though this method is actually only in one case called before the group is executed.
- successful case: explicitly specified to which group the annotation belongs via `groups=`.
- 1st unsuccessful case: nothing additional specifed, though groups should be inherited by the `@Test` annotation of the containing class.
- 2nd unsuccessful case: `alwaysRun=true` specified
  There is a <a href="https://github.com/quandor/testng-issue-118/">maven project</a> that reflects the described behaviour.

**Comment by paulwinters** [comment]
@cbeust 

This still is an issue with the latest build of TestnG 6.9.10 with @BeforeGroups not inheriting group sets.

A more detailed break down is found here.
http://stackoverflow.com/questions/35495937/testng-beforegroupsinheritgroups-true-usage

**Comment by krmahadevan** [comment]
Here's my observations ( This is as of TestNG `7.0.0-beta1` )

* `alwasyRun` is **NOT** applicable for `@BeforeGroups` . The javadoc clearly says that.
* With respect to  the attribute `inherited=true` in `@BeforeGroups`, yes its not being honored and this looks like a bug.

The below sample will reproduce the problem

```java
import org.testng.Assert;
import org.testng.annotations.BeforeGroups;
import org.testng.annotations.Test;

@Test(groups = "group1")
public class Group1Test {
   private Object testObject;

   @BeforeGroups
   public void setUpGroup() {
      testObject = new Object();
   }

   public void test1() {
      Assert.assertNotNull(testObject, "@BeforeGroups not invoked if nothing explicitly specified");
   }
}
```

testng xml

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd">
<suite name="Default suite">

  <test verbose="0" name="group1 included">
    <groups>
      <run>
        <include name="group1"/>
      </run>
    </groups>
    <classes>
      <class name="testng_issue_118.Group1Test"/>
    </classes>
  </test>
</suite>

```

**Comment by petar-lichev** [comment]
@krmahadevan Doesn't this break the @BeforeClass and @AfterClass logic? 
I have a class which contains a method annotated with @BeforeClass and with the change here

`case AFTER_GROUPS:
          afterGroups = configuration.getAfterGroups();
          create = shouldCreateBeforeAfterGroup(afterGroups, annotationFinder, clazz, configuration.getInheritGroups());
          isBeforeTestMethod = true;
          break;`

create becomes true, while previously it was false. This broke our build because a circular dependency was found due to this change. Is it expected that the methods annotated with @BeforeClass are collected and ordered as AfterGroups methods? (I have no previous experience with TestNG)

**Comment by krmahadevan** [comment]
@petar-lichev - do you have a sample project to share so that I can use it to reproduce the issue that you are experiencing?

**Comment by petar-lichev** [comment]
@krmahadevan Yes, I pushed it here - https://github.com/petar-lichev/sample-project
There are 2 branches - master and 6.14.3. 

master branch uses testng version 7.4.0. When you run the SampleTest you get:

`org.testng.TestNGException: 
The following methods have cyclic dependencies:
BaseClass.setupMethod2()[pri:0, instance:null]
SampleTest.setupMethod3()[pri:0, instance:null]
BaseClass.setupMethod1()[pri:0, instance:null]`

Which seems to come from _getAfterGroupsConfigurationMethods_ and the change in _findConfiguration_
where 

`shouldCreateBeforeAfterGroup(afterGroups, annotationFinder, clazz, configuration.getInheritGroups());`

returns true, because of the group configuration of the class level annotation of SampleTest. Which itself results in some cyclic dependency between
**SampleTest.setupMethod3** -> BaseClass.setupMethod2 -> BaseClass.setupMethod1 -> **SampleTest.setupMethod3**

I've also pushed the branch 6.14.3(test passes there) to showcase the different behaviour in_getAfterGroupsConfigurationMethods_ . 

Any idea if that is expected and it was the intended change in behaviour?

**Comment by krmahadevan** [comment]
@petar-lichev 

> There are 2 branches - master and 6.14.3. master branch uses testng version 7.4.0. 

The latest released versions are:

* `7.5.1` - Last supported version if you are still on JDK8. No bug fixes will be done on top of this version.
* `7.8.0` - Last released version that needs atleast JDK11. All bug fixes, feature requests etc are done on top of this.

Can you please help upgrade to `7.8.0` and update your findings ?

**Comment by petar-lichev** [comment]
@krmahadevan Same error. testMethodFinder.getAfterGroupsConfigurationMethods(m_testClass) throws the Exception about a circular dependency between the methods. 

![image](https://github.com/testng-team/testng/assets/41376193/de5d2a06-60ed-49e8-aa67-67e9406fb2a1)

**Comment by krmahadevan** [comment]
Will take a look. Thanks.

**Comment by petar-lichev** [comment]
`  private static boolean shouldCreateBeforeAfterGroup(
      String[] groups, IAnnotationFinder finder, Class<?> clazz, boolean isInheritGroups) {
    if (!isInheritGroups) {
      return groups.length > 0;
    }
    ITestAnnotation test = AnnotationHelper.findTest(finder, clazz);
    if (test == null) {
      return groups.length > 0;
    }
    return groups.length > 0 || test.getGroups().length > 0;
  }`


Basically with the adding of the shouldCreateBeforeAfterGroup method in the switch case, the behaviour changed, because of the @ Test(**groups**={"some_group"}) and the following now returns true: 

`test.getGroups().length > 0`

![image](https://github.com/testng-team/testng/assets/41376193/28d1e8c6-bef5-4e68-a258-7ec0bdb50434)

**Comment by petar-lichev** [comment]
@krmahadevan Hi, did you manage to take a look?
